transport_plugins/websocket/iotile_transport_websocket/device_adapter.py

```diff
@@ -0,0 +1,85 @@
+from iotile.core.exceptions import ArgumentError, HardwareError, TimeoutExpiredError
+from connection_manager import ConnectionManager
+
+
+class WebSocketDeviceAdapter(DeviceAdapter):
+    """ A device adapter allowing connections to devices over websockets
+
+    Args:
+        url (string): A url for the websocket server in form of server:port/path
+    """
+
+    def __init__(self, url):
+        super(WebSocketDeviceAdapter, self).__init__()
+
+        self.set_config('default_timeout', 10.0)
+        self.set_config('expiration_time', 60.0)
+        self.set_config('probe_supported', True)
+        self.set_config('probe_required', True)
+
+        path = "ws://{0}".format(url)
+        self.client = ValidatingWSClient(path)
+        self.client.start()
+
+        self.connections = ConnectionManager(self.id)
+        self.connections.start()
+
+    def connect_async(self, connection_id, connection_string, callback):
+        print(connection_id, connection_string)
+        context = {}
+        self.connections.begin_connection(connection_id, connection_string, callback, context, self.get_config('default_timeout'))
+
+        result = self.client.send_command('connect', {})
+        print(result)
+        # FIXME: async after creating WebSocketHandler2
+        self.connections.finish_connection(connection_id, result['success'], result.get('reason', None))
+
+    def disconnect_async(self, connection_id, callback):
+        try:
+            self.connections.get_connection_id(connection_id)
+        except ArgumentError:
+            callback(connection_id, self.id, False, "Could not find connection information")
+            return
+
+        self.connections.begin_disconnection(connection_id, callback, self.get_config('default_timeout'))
+
+        result = self.client.send_command('disconnect', {})
+        print(result)
+        # FIXME: async after creating WebSocketHandler2
+        self.connections.finish_disconnection(connection_id, result['success'], result.get('reason', None))
+
+    def probe_async(self, callback):
+        result = self.client.send_command('scan', {})
+        if not result['success']:
+            callback(self.id, False, 'Error while scanning through ws : {}'.format(result['reason']))
+        else:
+            # FIXME: async after creating WebSocketHandler2
+            for dev in result['devices']:
+                self._trigger_callback('on_scan', self.id, dev, self.get_config('expiration_time'))
+            callback(self.id, True, None)
+
+    def stop_sync(self):
+        connections = self.connections.get_connections()
+
+        for connection in connections:
+            try:
+                self.disconnect_sync(connection)
+            except HardwareError:
+                pass
+
+        self.client.stop()
+        self.connections.stop()
+
+    def periodic_callback(self):
+        try:
+            self.client.send_ping()
```

Comment:
I appreciate the intent here but I would remove this for now and if you want to add it back in, it will need to be nonblocking since periodic callback for device adapters in the `iotile-gateway` will be called in the main event loop as a periodic callback so they can't block the entire loop with a possible 10 second timeout. 

I wonder if this is necessary since the websocket client should flag us on disconnection anyway since it knows when the tcp socket goes away.  If you feel it is necessary then it would need to be done in a nonblocking way where, for example, you check when the last time you heard from the remote server was and if it's been longer than a fixed time you send a ping and set a flag that you have pinged and then the next periodic_callback if you have sent a ping and still haven't heard back you disconnect and stop.

Reply:
Indeed, the `closed` function in `ValidatingWebSocket` is called if the server unexpectedly disconnect but there is nothing triggered to display something to the user as this `closed` function is usually used to wait for a normal disconnection...

I remove my code in `periodic_callback` because indeed it is not async(I forgot to add a FIXME there ^^), but I think we should handle the case where we are unexpectedly disconnected from the gateway, to make it close properly the connections and to warn the user